scipy/sparse/csgraph/_flow.pyx

```diff
@@ -0,0 +1,395 @@
+    a : csr_matrix
+        The square matrix in CSR format representing a directed graph
+
+    Returns
+    -------
+    res : csr_matrix
+        A new matrix in CSR format in which the missing edges are represented
+        by explicit zeros.
+
+    """
+    # This is equivalent to just looping over all (i, j) and letting
+    # the weight of (j, i) be zero if it is not already defined.
+    # Using NumPy to do so is several orders of magnitude faster
+    # than the naive solution. We first create create the sorted
+    # list {i*n + j} for nonzero (i, j), then use binary search
+    # to determine if j*n + i belongs to the list. To ensure that
+    # the haystack below is sorted, we presort the column indices if
+    # necessary.
+    if not a.has_sorted_indices:
+        a = a.sorted_indices()
+    acoo = a.tocoo()
+    rows = acoo.row
+    cols = acoo.col
+    n = a.shape[0]
+    haystack = rows*n + cols
+    needles = cols*n + rows
+    indices = np.searchsorted(haystack, needles)
+    # As searchsorted returns a.nnz when a given needle is
+    # larger than the largest element of the haystack, we
+    # explicitly add an element not in needles at the end.
+    haystack = np.concatenate([haystack, [-1]])
+    indexed = haystack[indices]
+    diff = indexed != needles
+    # At this point, diff contains the indices of the (j, i)
+    # to be added. We make use of the COO initializer for CSR
+    # matrices which ensures that 0s are added explicitly.
+    indices_to_add = np.where(diff)[0]
+    data_to_add = np.zeros((indices_to_add.shape[0],), dtype=np.int32)
+    acoo.row = np.concatenate([acoo.row, cols[indices_to_add]])
+    acoo.col = np.concatenate([acoo.col, rows[indices_to_add]])
+    acoo.data = np.concatenate([acoo.data, data_to_add])
+    # Repeat process to get reverse pointers
+    return acoo.tocsr()
+
+
+def _make_edge_pointers(a):
+    """Create for each edge pointers to its reverse and its tail."""
+    # Just as above, one way to match (i, j) with (j, i) is to match
+    # i*n + j with j*n + i, which we can do by ensuring that we
+    # have a sorted collection of {i*n + j} and using binary search
+    # to find j*n + i here.
+    n = a.shape[0]
+    acoo = a.tocoo()
+    rows = acoo.row
+    cols = acoo.col
+    haystack = rows*n + cols
+    needles = cols*n + rows
+    rev_edge_ptr = np.searchsorted(haystack, needles).astype(np.int32)
+    return rev_edge_ptr, rows
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef int[:] _edmonds_karp(
+        int[:] edge_ptr,
+        int[:] tails,
+        int[:] heads,
+        int[:] capacities,
+        int[:] rev_edge_ptr,
+        int source,
+        int sink):
+    """Solves the maximum flow problem using the Edmonds--Karp algorithm.
+
+    This assumes that for every edge in the graph, the edge in the opposite
+    direction is also in the graph (possibly with capacity 0).
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length |V| + 1
```

Comment:
just a note on the `| |` (abs) symbols: those have a special meaning in Sphinx, and they shouldn't normally be used (it can break the doc build). I think it's not a problem here because it's in a private function so the docstring doesn't get parsed. But better to always enclose this in double backticks.

Reply:
Gotcha. Then I think I'll opt to enclose them in `:math:` so that in case they do get parsed, they'll look good as well.